File: app/src/test/java/com/soundcloud/android/playlists/PostsStorageTest.java
package com.soundcloud.android.playlists;

import static com.soundcloud.android.storage.TableColumns.Activities.SOUND_ID;
import static com.soundcloud.android.storage.TableColumns.Activities.SOUND_TYPE;
import static com.soundcloud.android.storage.Tables.Sounds.TYPE_PLAYLIST;
import static com.soundcloud.propeller.query.Query.from;

import com.soundcloud.android.api.model.ApiPlaylist;
import com.soundcloud.android.api.model.ApiTrack;
import com.soundcloud.android.model.Association;
import com.soundcloud.android.model.Urn;
import com.soundcloud.android.storage.Table;
import com.soundcloud.android.storage.TableColumns;
import com.soundcloud.android.sync.activities.ApiPlaylistRepostActivity;
import com.soundcloud.android.testsupport.StorageIntegrationTest;
import com.soundcloud.android.testsupport.fixtures.ModelFixtures;
import com.soundcloud.android.utils.TestDateProvider;
import com.soundcloud.propeller.query.Query;
import com.soundcloud.propeller.test.assertions.QueryAssertions;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;

import java.util.Arrays;
import java.util.Collections;
import java.util.Date;

public class PostsStorageTest extends StorageIntegrationTest {

    private static final Date POSTED_DATE_1 = new Date(100000);
    private static final Date POSTED_DATE_2 = new Date(200000);
    private static final Date POSTED_DATE_3 = new Date(300000);

    private PostsStorage storage;

    @Mock private RemovePlaylistCommand removePlaylistCommand;

    @Before
    public void setUp() throws Exception {
        storage = new PostsStorage(propellerRxV2(),
                                   new TestDateProvider(),
                                   removePlaylistCommand);
    }

    @Test
    public void shouldLoadAllPlaylistPosts() throws Exception {
        final PlaylistAssociation playlist1 = createPlaylistPostAt(POSTED_DATE_1);
        final PlaylistAssociation playlist2 = createPlaylistPostAt(POSTED_DATE_2);

        storage.loadPostedPlaylists(10, Long.MAX_VALUE).test()
               .assertValue(Arrays.asList(playlist2.getAssociation(), playlist1.getAssociation()));
    }

    @Test
    public void shouldAdhereToLimit() throws Exception {
        createPlaylistPostAt(POSTED_DATE_1);
        final PlaylistAssociation playlist2 = createPlaylistPostAt(POSTED_DATE_2);

        storage.loadPostedPlaylists(1, Long.MAX_VALUE).test()
               .assertValue(Collections.singletonList(playlist2.getAssociation()));
    }

    @Test
    public void shouldAdhereToPostedTime() throws Exception {
        final PlaylistAssociation playlist1 = createPlaylistAssociation(POSTED_DATE_1, POSTED_DATE_1);
        createPlaylistAssociation(POSTED_DATE_3, POSTED_DATE_1);

        storage.loadPostedPlaylists(2, POSTED_DATE_2.getTime()).test()
               .assertValue(Collections.singletonList(playlist1.getAssociation()));
    }

    @Test
    public void shouldNotIncludePlaylistsNotPresentInTheCollectionItemsTable() throws Exception {
        final PlaylistAssociation playlist1 = createPlaylistPostAt(POSTED_DATE_1);
        final PlaylistAssociation playlist2 = createPlaylistPostAt(POSTED_DATE_2);
        createPlaylistAt(POSTED_DATE_3); // deleted

        storage.loadPostedPlaylists(10, Long.MAX_VALUE).test()
               .assertValue(Arrays.asList(playlist2.getAssociation(), playlist1.getAssociation()));
    }

    @Test
    public void shouldNotConfuseTracksForPlaylists() throws Exception {
        final PlaylistAssociation playlist1 = createPlaylistPostAt(POSTED_DATE_1);
        final PlaylistAssociation playlist2 = createPlaylistPostAt(POSTED_DATE_2);

        createTrackPostWithId(playlist2.getPlaylist().urn().getNumericId());

        storage.loadPostedPlaylists(10, Long.MAX_VALUE).test()
               .assertValue(Arrays.asList(playlist2.getAssociation(), playlist1.getAssociation()));
    }

    @Test
    public void removeAssociatedActivitiesWhenMarkingPlaylistPendingRemovals() {
        final ApiPlaylist playlist = ModelFixtures.create(ApiPlaylist.class);
        final ApiPlaylistRepostActivity apiActivityItem = ModelFixtures.apiPlaylistRepostActivity(playlist);
        testFixtures().insertPlaylistRepostActivity(apiActivityItem);

        storage.markPendingRemoval(playlist.getUrn()).subscribe();

        final Query query = from(Table.Activities)
                .whereEq(SOUND_ID, playlist.getId())
                .whereEq(SOUND_TYPE, TYPE_PLAYLIST);

        QueryAssertions.assertThat(select(query)).isEmpty();
    }

    @Test
    public void removedSoundStreamEntryAssociatedWithRemovedPlaylist() {
        final ApiPlaylist playlist = ModelFixtures.create(ApiPlaylist.class);
        testFixtures().insertStreamPlaylistPost(playlist.getId(), 123L);

        storage.markPendingRemoval(playlist.getUrn()).subscribe();

        final Query query = from(Table.SoundStream)
                .whereEq(TableColumns.SoundStream.SOUND_ID, playlist.getId())
                .whereEq(TableColumns.SoundStream.SOUND_TYPE, TYPE_PLAYLIST);

        QueryAssertions.assertThat(select(query)).isEmpty();
    }

    @Test
    public void removedPostFromRemovedPlaylist() {
        final PlaylistAssociation playlist1 = createPlaylistPostAt(POSTED_DATE_1);
        final PlaylistAssociation playlist2 = createPlaylistPostAt(POSTED_DATE_2);


        storage.markPendingRemoval(playlist1.getTargetUrn()).subscribe();

        storage.loadPostedPlaylists(10, Long.MAX_VALUE).test()
               .assertValue(Arrays.asList(playlist2.getAssociation()));
    }

    private PlaylistAssociation createPlaylistPostAt(Date playlistPostedAt) {
        return createPlaylistAssociation(playlistPostedAt, playlistPostedAt);
    }

    private PlaylistAssociation createPlaylistAssociation(Date postedAt, Date playlistCreatedAt) {
        final ApiPlaylist apiPlaylist = createPlaylistAt(playlistCreatedAt);
        final Playlist.Builder playlist = ModelFixtures.playlistBuilder(apiPlaylist)
                                                       .isLikedByCurrentUser(false)
                                                       .isRepostedByCurrentUser(false)
                                                       .isMarkedForOffline(false);

        insertPlaylistPost(apiPlaylist.getUrn().getNumericId(), postedAt);
        return createPlaylistAssociation(playlist, postedAt);
    }

    private PlaylistAssociation createPlaylistAssociation(Playlist.Builder builder, Date createdAt) {
        final Playlist playlist = builder.isLikedByCurrentUser(false).build();
        return PlaylistAssociation.create(playlist, new Association(playlist.urn(), createdAt));
    }

    private ApiPlaylist createPlaylistAt(Date creationDate) {
        return testFixtures().insertPlaylistWithCreatedAt(creationDate);
    }

    private void insertPlaylistPost(long playlistId, Date postedAt) {
        testFixtures().insertPlaylistPost(playlistId, postedAt.getTime(), false);
    }

    private void createTrackPostWithId(long trackId) {
        ApiTrack apiTrack = ModelFixtures.create(ApiTrack.class);
        apiTrack.setUrn(Urn.forTrack(trackId));
        testFixtures().insertTrack(apiTrack);
        testFixtures().insertTrackPost(apiTrack.getId(), apiTrack.getCreatedAt().getTime(), false);
    }
}

File: app/src/test/java/com/soundcloud/android/posts/PostsStorageTest.kt
package com.soundcloud.android.posts

import com.nhaarman.mockito_kotlin.whenever
import com.soundcloud.android.accounts.AccountOperations
import com.soundcloud.android.api.model.ApiPlaylist
import com.soundcloud.android.api.model.ApiTrack
import com.soundcloud.android.api.model.ApiUser
import com.soundcloud.android.model.Association
import com.soundcloud.android.model.Urn
import com.soundcloud.android.playlists.Playlist
import com.soundcloud.android.playlists.PlaylistAssociation
import com.soundcloud.android.playlists.RemovePlaylistCommand
import com.soundcloud.android.storage.Table
import com.soundcloud.android.storage.TableColumns
import com.soundcloud.android.storage.Tables
import com.soundcloud.android.testsupport.PlaylistFixtures
import com.soundcloud.android.testsupport.StorageIntegrationTest
import com.soundcloud.android.testsupport.TrackFixtures
import com.soundcloud.android.testsupport.fixtures.ModelFixtures
import com.soundcloud.android.utils.TestDateProvider
import com.soundcloud.propeller.query.Query.from
import com.soundcloud.propeller.test.assertions.QueryAssertions
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import java.util.Arrays
import java.util.Date

@Suppress("IllegalIdentifier")
class PostsStorageTest : StorageIntegrationTest() {

    private lateinit var storage: PostsStorage
    private lateinit var user: ApiUser

    @Mock private lateinit var accountOperations: AccountOperations
    @Mock private lateinit var removePlaylistCommand: RemovePlaylistCommand

    @Before
    fun setUp() {
        user = testFixtures().insertUser()

        storage = PostsStorage(propellerRxV2(), TestDateProvider(), removePlaylistCommand)

        whenever(accountOperations.loggedInUserUrn).thenReturn(user.urn)
    }

    @Test
    @Throws(Exception::class)
    fun `loadPostedTracksSortedByDateDesc loads posted tracks in descending order`() {
        val lastPost = createTrackPostForLastPostedAt(POSTED_DATE_2)
        val firstPost = createTrackPostForLastPostedAt(POSTED_DATE_1)

        storage.loadPostedTracksSortedByDateDesc().test().assertValue(listOf(lastPost, firstPost))
    }

    @Test
    @Throws(Exception::class)
    fun `loadPostedPlaylists loads all posted playlists`() {
        val playlist1 = createPlaylistPostAt(POSTED_DATE_1)
        val playlist2 = createPlaylistPostAt(POSTED_DATE_2)

        storage.loadPostedPlaylists(10, java.lang.Long.MAX_VALUE).test()
                .assertValue(Arrays.asList(playlist2.association, playlist1.association))
    }

    @Test
    @Throws(Exception::class)
    fun `loadPostedPlaylists loads a limited number of posted playists`() {
        createPlaylistPostAt(POSTED_DATE_1)
        val playlist2 = createPlaylistPostAt(POSTED_DATE_2)

        storage.loadPostedPlaylists(1, java.lang.Long.MAX_VALUE).test()
                .assertValue(listOf(playlist2.association))
    }

    @Test
    @Throws(Exception::class)
    fun `loadPostedPlaylists only returns playlists posted after the given date`() {
        val playlist1 = createPlaylistAssociation(POSTED_DATE_1, POSTED_DATE_1)
        createPlaylistAssociation(POSTED_DATE_3, POSTED_DATE_1)

        storage.loadPostedPlaylists(2, POSTED_DATE_2.time).test()
                .assertValue(listOf(playlist1.association))
    }

    @Test
    @Throws(Exception::class)
    fun `loadPostedPlaylists only loads playlists that are in the CollectionItems table`() {
        val playlist1 = createPlaylistPostAt(POSTED_DATE_1)
        val playlist2 = createPlaylistPostAt(POSTED_DATE_2)
        createPlaylistAt(POSTED_DATE_3) // deleted

        storage.loadPostedPlaylists(10, java.lang.Long.MAX_VALUE).test()
                .assertValue(Arrays.asList(playlist2.association, playlist1.association))
    }

    @Test
    @Throws(Exception::class)
    fun `loadPostedPlaylists only loads returns playlists`() {
        val playlist1 = createPlaylistPostAt(POSTED_DATE_1)
        val playlist2 = createPlaylistPostAt(POSTED_DATE_2)

        createTrackPostWithId(playlist2.playlist.urn().numericId)

        storage.loadPostedPlaylists(10, java.lang.Long.MAX_VALUE).test()
                .assertValue(Arrays.asList(playlist2.association, playlist1.association))
    }

    @Test
    fun `markPlaylistPendingRemoval removes associated entry from Activities View`() {
        val playlist = PlaylistFixtures.apiPlaylist()
        val apiActivityItem = ModelFixtures.apiPlaylistRepostActivity(playlist)
        testFixtures().insertPlaylistRepostActivity(apiActivityItem)

        storage.markPlaylistPendingRemoval(playlist.urn).subscribe()

        val query = from(Table.Activities)
                .whereEq(TableColumns.Activities.SOUND_ID, playlist.id)
                .whereEq(TableColumns.Activities.SOUND_TYPE, com.soundcloud.android.storage.Tables.Sounds.TYPE_PLAYLIST)

        QueryAssertions.assertThat(select(query)).isEmpty
    }

    @Test
    fun `markPlaylistPendingRemoval removes associated entry from SoundStream View`() {
        val playlist = PlaylistFixtures.apiPlaylist()
        testFixtures().insertStreamPlaylistPost(playlist.id, 123L)

        storage.markPlaylistPendingRemoval(playlist.urn).subscribe()

        val query = from(Table.SoundStream)
                .whereEq(TableColumns.SoundStream.SOUND_ID, playlist.id)
                .whereEq(TableColumns.SoundStream.SOUND_TYPE, com.soundcloud.android.storage.Tables.Sounds.TYPE_PLAYLIST)

        QueryAssertions.assertThat(select(query)).isEmpty
    }

    @Test
    fun `markPlaylistPendingRemoval removes associated entry from Posts table`() {
        val playlist1 = createPlaylistPostAt(POSTED_DATE_1)
        val playlist2 = createPlaylistPostAt(POSTED_DATE_2)

        storage.markPlaylistPendingRemoval(playlist1.targetUrn).subscribe()

        storage.loadPostedPlaylists(10, java.lang.Long.MAX_VALUE).test()
                .assertValue(Arrays.asList(playlist2.association))
    }

    @Test
    fun `markPlaylistPendingRemoval removes associated entry from Sounds table`() {
        val playlist = PlaylistFixtures.apiPlaylist()
        testFixtures().insertStreamPlaylistPost(playlist.id, 123L)

        storage.markPlaylistPendingRemoval(playlist.urn).subscribe()

        val query = from(Tables.Sounds.TABLE)
                .whereEq(Tables.Sounds._ID, playlist.id)
                .whereEq(Tables.Sounds._TYPE, com.soundcloud.android.storage.Tables.Sounds.TYPE_PLAYLIST)

        QueryAssertions.assertThat(select(query)).isEmpty
    }

    @Test(expected = IllegalArgumentException::class)
    fun `markPlaylistPendingRemoval throws an exception if urn is not a playlist urn`() {
        storage.markPlaylistPendingRemoval(Urn.forTrack(123L)).subscribe()
    }

    @Test(expected = IllegalArgumentException::class)
    fun `removePlaylist throws an exception if urn is not a playlist urn`() {
        storage.removePlaylist(Urn.forTrack(123L)).subscribe()
    }

    private fun createPlaylistAssociation(postedAt: Date, playlistCreatedAt: Date): PlaylistAssociation {
        val apiPlaylist = createPlaylistAt(playlistCreatedAt)
        val playlist = PlaylistFixtures.playlistBuilder(apiPlaylist)
                .isLikedByCurrentUser(false)
                .isRepostedByCurrentUser(false)

        insertPlaylistPost(apiPlaylist.urn.numericId, postedAt)
        return createPlaylistAssociation(playlist, postedAt)
    }

    private fun createPlaylistPostAt(playlistPostedAt: Date): PlaylistAssociation = createPlaylistAssociation(playlistPostedAt, playlistPostedAt)

    private fun createPlaylistAssociation(builder: Playlist.Builder, createdAt: Date): PlaylistAssociation {
        val playlist = builder.isLikedByCurrentUser(false).build()
        return PlaylistAssociation.create(playlist, Association(playlist.urn(), createdAt))
    }

    private fun createPlaylistAt(creationDate: Date): ApiPlaylist = testFixtures().insertPlaylistWithCreatedAt(creationDate)

    private fun insertPlaylistPost(playlistId: Long, postedAt: Date) {
        testFixtures().insertPlaylistPost(playlistId, postedAt.time, false)
    }

    private fun createTrackPostWithId(trackId: Long) {
        val apiTrack = TrackFixtures.apiTrack(Urn.forTrack(trackId))
        testFixtures().insertTrack(apiTrack)
        testFixtures().insertTrackPost(apiTrack.id, apiTrack.createdAt.time, false)
    }

    private fun createTrackAt(creationDate: Date): ApiTrack = testFixtures().insertTrackWithCreationDate(user, creationDate)

    private fun createTrackPostWithId(trackId: Long, postedAt: Date) {
        testFixtures().insertTrackPost(trackId, postedAt.time, false)
    }

    private fun createTrackPostForLastPostedAt(postedAt: Date): Association {
        val track = createTrackAt(postedAt)
        createTrackPostWithId(track.urn.numericId, postedAt)
        return Association(track.urn, postedAt)
    }

    companion object {
        private val POSTED_DATE_1 = Date(100000)
        private val POSTED_DATE_2 = Date(200000)
        private val POSTED_DATE_3 = Date(300000)
    }

}
